### DISCUSSION TITLE
Which protocol does ClickHouse.Client use behind the scene

### DISCUSSION THREAD

**Discussion by Ceyword** [DarkWanderer/ClickHouse.Client#348] [General]
Which protocol does ClickHouse.Client use behind the scene?

Is it HTTP?

Are we limited to HTTP, since it has limited functionality?

**Comment by DarkWanderer** [comment]
Hi @Ceyword 

ClickHouse.Client uses HTTP behind the scenes, yes. The main reason is at the start of development Native protocol was poorly documented; also, HTTP protocol has support for Kerberos authentication which native client lacks

Which features do you find missing which you think may be available in other protocols?

**Reply by Ceyword** [reply]
Hello @DarkWanderer ,
Thank you for your prompt response.
I am actually new to ClickHouse and researching to settle for the best options available.

My question is a reaction to the documentation on https://clickhouse.com/docs/en/interfaces/http :
> The HTTP interface lets you use ClickHouse on any platform from any programming language in a form of REST API. The HTTP interface is more limited than the native interface, but it has better language support.

Other follow-up questions:
1- Is ClickHouse.Client used in production?
2- What is your experience/knowledge about it's performance in production?
3- Any know bottle-neck/limitations at scale?
4- Does it have robustness features like automatic retry etc.?

Talking of features I will like, I am excited about full-text search capability using inverted indexes. This is still experimental but it will be nice to have it as soon as possible.

Thank you.

**Reply by DarkWanderer** [reply]
1. Yes, by multiple users
2. The performance is high - it is possible to achieve better, but at the cost of functionality
3. Watch the number concurrent connections when working with same HTTP server
4. No, and it is not the correct place to have them. It is customizable enough though for you to implement it yourself: https://stackoverflow.com/questions/19260060/retrying-httpclient-unsuccessful-requests

Indexes are a question for developers database itself, not the client library
